include/rocksdb/trace_record.h

```diff
@@ -0,0 +1,133 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+// To do: ROCKSDB_LITE ?
+
+#include "rocksdb/db.h"
+#include "rocksdb/slice.h"
+#include "rocksdb/write_batch.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+// Supported Trace types.
+enum TraceType : char {
+  kTraceNone = 0,
+  kTraceBegin = 1,
+  kTraceEnd = 2,
+  kTraceWrite = 3,
+  kTraceGet = 4,
+  kTraceIteratorSeek = 5,
+  kTraceIteratorSeekForPrev = 6,
+  // Block cache related types.
+  kBlockTraceIndexBlock = 7,
+  kBlockTraceFilterBlock = 8,
+  kBlockTraceDataBlock = 9,
+  kBlockTraceUncompressionDictBlock = 10,
+  kBlockTraceRangeDeletionBlock = 11,
+  // For IOTracing.
+  kIOTracer = 12,
+  // For query tracing
+  kTraceMultiGet = 13,
+  // All trace types should be added before kTraceMax
+  kTraceMax,
+};
+
+// Base class for all types of trace records.
+class TraceRecord {
```

Comment:
This class appears to be a more complex version of Trace (from trace_replay.h).  Is there a reason not to simply move Trace into the public API?  There is already a "TODO" comment there stating that should be done (probably renamed to "class TraceRecord").  You could create accessor functions for the various payload types in the TraceRecord and do away with the many types of classes (e.g. 

Status TraceRecord::GetWriteBatch(WriteBatch *wb) const { 
  if (payload_map & kWriteBatchData) { /
    / fill and return ok;
  } else {
   return Status::NotFound("No WriteBatch in this record");
 } 
}

Reply:
I think we are not going to make Trace public, so payload content will not be visible to the external. That is to say, TraceRecord should not contain anything from payload. There won't be public APIs that convert a Trace to a TraceRecord, or from payload to key(s)/cf_id(s), etc. That also the reason why TraceRecord is a class type, not struct type, since you cannot have a struct that has all information for different operations.

Reply:
I would argue that you are effectively making Trace public via the TraceRecord, but in a way that is more complex than the original Trace (lots of classes and potential to go astray).  I would suggest making Trace public and have a way of getting the fields out of it that you want/need.  With some thought you may not need to make the Payload types public and can get/set them via a public API (or via Trace constructors).

There is quite a bit to this TraceRecord class hierarchy and I think most of it duplicates something that already exists in the Trace/Payload methods.  Why re-invent the wheel?

Reply:
We can make Trace public, but then we also have to make `TracePayloadType`, `WritePayload`, `GetPayload`, `IterPayload`, `MultiGetPayload`, `TraceHelper`, and other types of payloads used in block cache tracing and IO tracing public. These payload types are specific to the trace type thus are not generic. This incurs more complexity than having a single `TraceRecord` class.

Reply:
I do not think you need to make TracePayloadType, WritePayload, etc public.  

If you wanted to only read (and not create) new Trace records, you need to know the TraceType and have accessor methods to read the payload fields (GetWriteBatch, GetColumnFamilyID, etc).  

If you want to be able to create new Trace records, you could either do that via various constructors (mimicking the Payload arguments) or have setter functions (potentially more complex).

Reply:
Yes, if we don't make payload types public, we need to make TraceHelper public, or provide another class with lots of static functions to GetGetTraceKey(), GetGetTraceCFID(), GetWriteTraceRep(), ... etc.

Later whenever we add a new trace type, we need add one or multiple public functions to this class. Isn't it more complex than using `TraceRecord`?

Reply:
I do not think it is more complex and TraceHelper does not need to be public.  GetTraceKey would internally use TraceHelper but TraceHelper does not need to be public.

You only need different methods for different types of data.  If multiple types return a ColumnFamilyId, they all can use the same method.

The proposed approach requires a new class for every type.  The approach I am suggesting requires a new method for every unique field in payload (from the payload mask).   If a new Trace type is added and we are allowing users to construct them, then a new constructor would also be required.